common/lib/core-interfaces/test-d/index.test-d.ts

```diff
@@ -0,0 +1,88 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+/* eslint-disable unicorn/filename-case */
+// eslint-disable-next-line import/no-extraneous-dependencies
+import {expectError} from "tsd";
+import { IFluidLoadable, IProvideFluidLoadable, FluidObject, FluidObjectKeys } from "../dist";
+
+
+declare function getUnknownFluidObject(): FluidObject;
+
+declare function useUnknownFluidObject(params: FluidObject | undefined): void;
+
+declare function useProvider<T extends FluidObject>(params: FluidObject<T> | undefined): void;
+
+declare function useProviderKey<T,TKey extends FluidObjectKeys<T> = FluidObjectKeys<T>>(key: TKey): void;
+
+declare function useLoadable(params: FluidObject<IFluidLoadable> | undefined): void;
+
+// test implicit conversions between FluidObject and a FluidObject with a provides interface
+{
+    const FluidObject: FluidObject<IProvideFluidLoadable> = getUnknownFluidObject();
+    useUnknownFluidObject(FluidObject);
+    useUnknownFluidObject(FluidObject.IFluidLoadable);
+    useProvider(FluidObject);
+    useProvider(FluidObject.IFluidLoadable);
+    useLoadable(FluidObject);
+    useLoadable(FluidObject.IFluidLoadable);
+    expectError(FluidObject.handle);
+    FluidObject.IFluidLoadable?.handle;
+    const unknown: FluidObject | undefined = FluidObject.IFluidLoadable;
+    useUnknownFluidObject(unknown);
+    useProvider(unknown);
+    useProvider<IFluidLoadable>(unknown);
+    useLoadable(unknown);
+}
+
+// test implicit conversions between FluidObject and a FluidObject with a implementation interface
+{
+    const foo: FluidObject<IFluidLoadable> = getUnknownFluidObject();
+    useUnknownFluidObject(foo);
+    useUnknownFluidObject(foo.IFluidLoadable);
+    useProvider(foo);
+    useProvider(foo.IFluidLoadable);
+    useLoadable(foo);
+    useLoadable(foo.IFluidLoadable);
+    expectError(foo.handle);
+    foo.IFluidLoadable?.handle;
+    const unknown: FluidObject | undefined = foo.IFluidLoadable;
+    useUnknownFluidObject(unknown);
+    useProvider(unknown);
+    useProvider<IFluidLoadable>(unknown);
+    useLoadable(unknown);
+}
+
+// test getting keys
+{
+    useProviderKey<IProvideFluidLoadable>(IFluidLoadable);
+    useProviderKey<IFluidLoadable>(IFluidLoadable);
+    const loadableKey: keyof IFluidLoadable = "handle";
+    expectError(useProviderKey<IFluidLoadable>(loadableKey));
+}
+
+// test implicit conversions between FluidObject and a FluidObject with a partial provider interface
+{
+
+    interface IProvideFoo{
+        IFoo: IFoo;
+    }
+    interface IFoo extends Partial<IProvideFoo>{
```

Comment:
@vladsud and @markfields i've added a new test which verifies that the new utility type, which i've renamed from `Provider` to `FluidObject` also works if the provider is partial/optional on the implementation. This means we only need to type to specify it, and we do not need to implementation to actually have it.

While not perfect, i do think this is closer to the pattern you also prefer where provider and implementor are decoupled.

also, @barnawhi. the partial thing may make functional objects work a bit better

Reply:
So when I try this in [TS Playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBANiA5jAvDUkRwKYDoGIDcAUMdgB4AOIATrFAJ6XYwBicArgJYAmA8gCMAVtmBQACjRAA3XthoBpbAwgAeACoAaGOskhKMClGxgeEGAGtlIAGY7Ul63fUA+VMRieY0GlzDIjEzMdPQMAfhgwbGl5GAAuSI4AWwFYwNNzXSlKCKiYmniYAHoiw3JgTh4WPyryeXMAbSsGOJ8-ZAAfMGTUmgBdBIBDMAYPLyz9MuMMxwZbHQaJyj7i0qWYLnNBmEps+UYYea0FpZXhnm11RdC+6+y+sa8vCPX04Oaju-1b09WywhgAEk9LIqqwQCBcIDwZDoRCoTDHk9kS9QkjkV44nl5OjMZFovISGQqLR6EwWOxuPxhKIoBoHBwwBYwCAAO5gNxocSDOhcQZwVTiLjACwabSU3iCERiEFyRTKNSuFwuImMZgwAByBIKaA+dgAFHCQDAOjBGVUbH5sDwAJSkIoAKgdHgdOgAFpszVAuHAuAc1SxgDRsINjFsYAGI27QxtzFA3SxGVxwIc7Py4LNzCn4yxECZ5MKI+SXTBWR7gG6NklKDgkiZ6Am2JxJTSxDtQ8YaGBcDAAAIQbAU5vU6VQF0l8EFCiDas4GAAAwlI9pShUqiNJqBACEeS556WQBw4DwYMGIEfYH4FwAiI3Xze3nc0a-z8cwB1FYnUOhF9VLqUrgqGicrM8z-q2dKuESNiMmIyZgBGNAMH6+oAI4JOBo42jAADeSIYNAMBJIMDCpE+GHDgBYjrk+IGoSQTwEPqxGkdgT5Qk+YS4AwNpEk8BGwCxqQwhRVJUXSRp0QxXhMUJ2AwgiEJceQvHEAAvqQFSDBA5hPlWNbYHWYBQOYgJ6XhTwtN4UC+P4DjXgAggAyqwAAi17SZ4Zk8gk3k6lGmwkBpxB+J2NiDMALDAlIoJsTyuFIn5vlPkFpChfI4WRdu8VvKZspVOZSJWW0-ipSFxkZRFUX5fJEIJU8Rq+TCZXpTQmVRTCUxBOY3K8vy641TCbgWV45AJN0KSEk86mkDZyFQPqOEwGNMAAIwAEwAMwwGpqlAA), it doesn't work. And that makes sense to me - your Keys type will be doing `keyof (IFoo | undefined)` (line 27) which is `never`.  Not sure why your tests pass -_-

Reply:
weird. not sure why there is an error there, but not the test, but regardless nice find. fix it with an exclude

Reply:
This works too:

```
TProp extends keyof Required<T>[TProp]
```

Maybe you don't have strict null checks enabled something weird like that?